File: VHKPlayer/VHKPlayer/Queries/QueryProcessor.cs
using System.Threading.Tasks;
using Autofac;
using VHKPlayer.Queries.Interfaces;

namespace VHKPlayer.Queries
{
    public class QueryProcessor : IQueryProcessor
    {
        private readonly IComponentContext _context;

        public QueryProcessor(IComponentContext context)
        {
            this._context = context;
        }

        public TResult Process<TResult>(IQuery<TResult> query)
        {
            var handlerType = typeof(IQueryHandler<,>).MakeGenericType(query.GetType(), typeof(TResult));
            dynamic handler = _context.Resolve(handlerType);

            return handler.Handle((dynamic) query);
        }

        public async Task<TResult> ProcessAsync<TResult>(IQuery<TResult> query)
        {
            return await Task.Factory.StartNew(() => Process(query)).ConfigureAwait(false);
        }
    }
}

File: core/src/main/java/com/vhkplayer/dao/queries/QueryProcessor.java
package com.vhkplayer.dao.queries;

import java.util.Map;
import java.util.Optional;

public class QueryProcessor {
    private final Map<Class, QueryHandlerCreationStrategy> handlerOptions;

    public QueryProcessor(Map<Class, QueryHandlerCreationStrategy> handlerOptions) {
        this.handlerOptions = handlerOptions;
    }

    public <TResult> Optional<TResult> process(Query<TResult> query) {
        QueryHandlerCreationStrategy<Query<TResult>, TResult> strategy = handlerOptions.get(query.getClass());
        if (strategy != null) {
            return Optional.of(strategy.createHandler().handle(query));
        }
        return Optional.empty();
    }
}
